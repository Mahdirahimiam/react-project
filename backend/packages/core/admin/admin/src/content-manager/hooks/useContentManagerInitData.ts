import { useEffect, useRef } from 'react';

import { useNotifyAT } from '@strapi/design-system';
import {
  Permission,
  hasPermissions,
  useFetchClient,
  useNotification,
  useRBACProvider,
  useStrapiApp,
} from '@strapi/helper-plugin';
import { Contracts } from '@strapi/plugin-content-manager/_internal/shared';
import { stringify } from 'qs';
import { useIntl } from 'react-intl';

import { HOOKS } from '../../constants';
import { useTypedDispatch, useTypedSelector } from '../../core/store/hooks';
import { getTranslation } from '../utils/translations';

const { MUTATE_COLLECTION_TYPES_LINKS, MUTATE_SINGLE_TYPES_LINKS } = HOOKS;

interface ContentManagerLink {
  permissions: Permission[];
  search: string | null;
  kind: string;
  title: string;
  to: string;
  uid: string;
  name: string;
  isDisplayed: boolean;
}

/**
 * TODO: refactor this whole thing.
 */
const useContentManagerInitData = () => {
  const dispatch = useTypedDispatch();
  const toggleNotification = useNotification();
  const state = useTypedSelector((state) => state['content-manager_app']);

  const { allPermissions } = useRBACProvider();
  const { runHookWaterfall } = useStrapiApp();
  const { notifyStatus } = useNotifyAT();
  const { formatMessage } = useIntl();
  const { get } = useFetchClient();

  const fetchDataRef = useRef(async () => {
    dispatch({
      type: 'ContentManager/App/GET_INIT_DATA',
    });

    try {
      const {
        data: {
          data: { components, contentTypes: models, fieldSizes },
        },
      } = await get<Contracts.Init.GetInitData.Response>('/content-manager/init');
      notifyStatus(
        formatMessage({
          id: getTranslation('App.schemas.data-loaded'),
          defaultMessage: 'The schemas have been successfully loaded.',
        })
      );

      const {
        data: { data: contentTypeConfigurations },
      } = await get<Contracts.ContentTypes.FindContentTypesSettings.Response>(
        '/content-manager/content-types-settings'
      );

      /**
       * We group these by the two types we support. We do with an object because we can use default
       * values of arrays to make sure we always have an array to manipulate further on if, for example,
       * a user has not made any single types.
       *
       * This means we have to manually add new content types to this hook if we add a new type â€“ but
       * the safety is worth it.
       */
      const { collectionType: collectionTypeLinks, singleType: singleTypeLinks } = models.reduce<{
        collectionType: Contracts.ContentTypes.ContentType[];
        singleType: Contracts.ContentTypes.ContentType[];
      }>(
        (acc, model) => {
          acc[model.kind].push(model);
          return acc;
        },
        {
          collectionType: [],
          singleType: [],
        }
      );

      const collectionTypeSectionLinks = generateLinks(
        collectionTypeLinks,
        'collectionTypes',
        contentTypeConfigurations
      );

      const singleTypeSectionLinks = generateLinks(singleTypeLinks, 'singleTypes');

      // Collection Types verifications
      const collectionTypeLinksPermissions = await Promise.all(
        collectionTypeSectionLinks.map(({ permissions }) =>
          hasPermissions(allPermissions, permissions)
        )
      );
      const authorizedCollectionTypeLinks = collectionTypeSectionLinks.filter(
        (_, index) => collectionTypeLinksPermissions[index]
      );

      // Single Types verifications
      const singleTypeLinksPermissions = await Promise.all(
        singleTypeSectionLinks.map(({ permissions }) => hasPermissions(allPermissions, permissions))
      );
      const authorizedSingleTypeLinks = singleTypeSectionLinks.filter(
        (_, index) => singleTypeLinksPermissions[index]
      );

      const { ctLinks } = runHookWaterfall(MUTATE_COLLECTION_TYPES_LINKS, {
        ctLinks: authorizedCollectionTypeLinks,
        models,
      });
      const { stLinks } = runHookWaterfall(MUTATE_SINGLE_TYPES_LINKS, {
        stLinks: authorizedSingleTypeLinks,
        models,
      });

      dispatch({
        type: 'ContentManager/App/SET_INIT_DATA',
        data: {
          authorizedCollectionTypeLinks: ctLinks,
          authorizedSingleTypeLinks: stLinks,
          contentTypeSchemas: models,
          components,
          fieldSizes,
        },
      });
    } catch (err) {
      console.error(err);

      toggleNotification({ type: 'warning', message: { id: 'notification.error' } });
    }
  });

  useEffect(() => {
    fetchDataRef.current();

    return () => {
      dispatch({
        type: 'ContentManager/App/RESET_INIT_DATA',
      });
    };
  }, [dispatch, toggleNotification]);

  return { ...state, refetchData: fetchDataRef.current };
};

const generateLinks = (
  links: Contracts.ContentTypes.ContentType[],
  type: 'collectionTypes' | 'singleTypes',
  configurations: Contracts.ContentTypes.FindContentTypesSettings.Response['data'] = []
) => {
  return links
    .filter((link) => link.isDisplayed)
    .map((link) => {
      const collectionTypesPermissions = [
        { action: 'plugin::content-manager.explorer.create', subject: link.uid },
        { action: 'plugin::content-manager.explorer.read', subject: link.uid },
      ];
      const singleTypesPermissions = [
        { action: 'plugin::content-manager.explorer.read', subject: link.uid },
      ];
      const permissions =
        type === 'collectionTypes' ? collectionTypesPermissions : singleTypesPermissions;

      const currentContentTypeConfig = configurations.find(({ uid }) => uid === link.uid);

      let search = null;

      if (currentContentTypeConfig) {
        const searchParams = {
          page: 1,
          pageSize: currentContentTypeConfig.settings.pageSize,
          sort: `${currentContentTypeConfig.settings.defaultSortBy}:${currentContentTypeConfig.settings.defaultSortOrder}`,
        };

        search = stringify(searchParams, { encode: false });
      }

      return {
        permissions,
        search,
        kind: link.kind,
        title: link.info.displayName,
        to: `/content-manager/${
          link.kind === 'collectionType' ? 'collection-types' : 'single-types'
        }/${link.uid}`,
        uid: link.uid,
        // Used for the list item key in the helper plugin
        name: link.uid,
        isDisplayed: link.isDisplayed,
      } satisfies ContentManagerLink;
    });
};

export { useContentManagerInitData };
export type { ContentManagerLink };
