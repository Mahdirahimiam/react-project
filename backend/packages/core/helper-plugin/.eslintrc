{
  "root": true,
  "extends": ["custom/front/typescript", "plugin:storybook/recommended"],
  "rules": {
    "import/no-unresolved": [
      "error",
      {
        // Eslint does not detect the package.json exports field
        // and reports broken imports for the following imports
        "ignore": ["@strapi/design-system/v2"]
      }
    ]
  },
  "overrides": [
    {
      "files": ["./jest.config.front.js", "./webpack.config.js"],
      "env": {
        "node": true
      },
      "rules": {
        "@typescript-eslint/no-var-requires": "off"
      }
    },
    {
      "files": ["./tests/*"],
      "env": {
        "jest": true
      }
    }
  ]
}
