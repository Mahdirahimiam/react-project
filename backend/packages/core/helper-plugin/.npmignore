############################
# dts
############################

dist/**/*.d.ts.map