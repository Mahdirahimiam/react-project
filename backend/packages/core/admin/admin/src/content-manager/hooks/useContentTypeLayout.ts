import { useCallback } from 'react';

import { useTypedSelector } from '../../core/store/hooks';

const useContentTypeLayout = () => {
  const currentLayout = useTypedSelector(
    (state) => state['content-manager_editViewLayoutManager'].currentLayout
  );

  const getComponentLayout = useCallback(
    (componentUid: string) => {
      return currentLayout?.components?.[componentUid] ?? {};
    },
    [currentLayout]
  );

  return { ...currentLayout, getComponentLayout };
};

export { useContentTypeLayout };
