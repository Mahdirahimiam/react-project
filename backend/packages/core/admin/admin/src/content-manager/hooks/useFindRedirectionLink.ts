import { useQueryParams } from '@strapi/helper-plugin';
import QueryString, { parse, stringify } from 'qs';

import { useTypedSelector } from '../../core/store/hooks';
import { isObject } from '../../utils/objects';

import type { ContentManagerAppState } from '../pages/App';

const useFindRedirectionLink = (slug: string) => {
  const [{ rawQuery }] = useQueryParams();
  const collectionTypesMenuLinks = useTypedSelector(
    (state) => state['content-manager_app'].collectionTypeLinks
  );
  const redirectionLink = getRedirectionLink(collectionTypesMenuLinks, slug, rawQuery);

  return redirectionLink;
};

/**
 * Updates the leafs of the first argument
 */
const mergeParams = (
  initialParams: QueryString.ParsedQs,
  params: QueryString.ParsedQs
): QueryString.ParsedQs => {
  return Object.keys(initialParams).reduce<QueryString.ParsedQs>((acc, current) => {
    const initialValue = initialParams[current];
    const nextValue = params[current] ?? initialValue;

    if (isObject(initialValue)) {
      // @ts-expect-error â€“ TODO: fix this type error.
      return { ...acc, [current]: mergeParams(initialValue, nextValue) };
    }

    acc[current] = nextValue;

    return acc;
  }, {});
};

const getRedirectionLink = (
  links: ContentManagerAppState['collectionTypeLinks'],
  slug: string,
  rawQuery: string
) => {
  const matchingLink = links.find(({ to }) => to.includes(slug));

  if (!matchingLink) {
    return '/';
  }

  const { to, search } = matchingLink;
  const searchQueryParams = parse(search ?? '');
  const currentQueryParams = parse(rawQuery.substring(1));

  const mergedParams = mergeParams(searchQueryParams, currentQueryParams);

  const link = `${to}?${stringify(mergedParams, { encode: false })}`;

  return link;
};

export { useFindRedirectionLink };
