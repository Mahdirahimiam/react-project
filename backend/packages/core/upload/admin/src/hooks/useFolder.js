import { useFetchClient, useNotification } from '@strapi/helper-plugin';
import { useQuery } from 'react-query';

import pluginId from '../pluginId';
import { getTrad } from '../utils';

export const useFolder = (id, { enabled = true } = {}) => {
  const toggleNotification = useNotification();
  const { get } = useFetchClient();

  const { data, error, isLoading } = useQuery(
    [pluginId, 'folder', id],
    async () => {
      const {
        data: { data },
      } = await get(`/upload/folders/${id}`, {
        params: {
          populate: {
            parent: {
              populate: {
                parent: '*',
              },
            },
          },
        },
      });

      return data;
    },
    {
      retry: false,
      enabled,
      staleTime: 0,
      cacheTime: 0,
      onError() {
        toggleNotification({
          type: 'warning',
          message: {
            id: getTrad('notification.warning.404'),
            defaultMessage: 'Not found',
          },
        });
      },
    }
  );

  return { data, error, isLoading };
};
