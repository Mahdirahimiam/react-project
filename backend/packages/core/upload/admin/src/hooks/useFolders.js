import * as React from 'react';

import { useNotifyAT } from '@strapi/design-system';
import { useFetchClient, useNotification } from '@strapi/helper-plugin';
import { stringify } from 'qs';
import { useIntl } from 'react-intl';
import { useQuery } from 'react-query';

import pluginId from '../pluginId';

export const useFolders = ({ enabled = true, query = {} } = {}) => {
  const { formatMessage } = useIntl();
  const toggleNotification = useNotification();
  const { notifyStatus } = useNotifyAT();
  const { folder, _q, ...paramsExceptFolderAndQ } = query;
  const { get } = useFetchClient();

  let params;

  if (_q) {
    params = {
      ...paramsExceptFolderAndQ,
      pagination: {
        pageSize: -1,
      },
      _q,
    };
  } else {
    params = {
      ...paramsExceptFolderAndQ,
      pagination: {
        pageSize: -1,
      },
      filters: {
        $and: [
          ...(paramsExceptFolderAndQ?.filters?.$and ?? []),
          {
            parent: {
              id: folder ?? {
                $null: true,
              },
            },
          },
        ],
      },
    };
  }

  const { data, error, isLoading } = useQuery(
    [pluginId, 'folders', stringify(params)],
    async () => {
      const {
        data: { data },
      } = await get('/upload/folders', { params });

      return data;
    },
    {
      enabled,
      staleTime: 0,
      cacheTime: 0,
      onError() {
        toggleNotification({
          type: 'warning',
          message: { id: 'notification.error' },
        });
      },
    }
  );

  React.useEffect(() => {
    if (data) {
      notifyStatus(
        formatMessage({
          id: 'list.asset.at.finished',
          defaultMessage: 'The folders have finished loading.',
        })
      );
    }
  }, [data, formatMessage, notifyStatus]);

  return { data, error, isLoading };
};
