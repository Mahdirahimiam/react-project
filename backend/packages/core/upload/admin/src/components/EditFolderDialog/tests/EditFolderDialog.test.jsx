import React from 'react';

import { lightTheme, ThemeProvider } from '@strapi/design-system';
import { NotificationsProvider } from '@strapi/helper-plugin';
import { within } from '@testing-library/dom';
import { act, fireEvent, render, screen, waitFor } from '@testing-library/react';
import { IntlProvider } from 'react-intl';
import { QueryClient, QueryClientProvider } from 'react-query';

import { useEditFolder } from '../../../hooks/useEditFolder';
import { useMediaLibraryPermissions } from '../../../hooks/useMediaLibraryPermissions';
import { EditFolderDialog } from '../EditFolderDialog';

jest.mock('@strapi/helper-plugin', () => ({
  ...jest.requireActual('@strapi/helper-plugin'),
  useQueryParams: jest.fn().mockReturnValue([{ query: {} }]),
  useFetchClient: jest.fn().mockReturnValue({
    put: jest.fn().mockImplementation({}),
  }),
}));

jest.mock('../../../hooks/useMediaLibraryPermissions');
jest.mock('../../../hooks/useFolderStructure');
jest.mock('../../../hooks/useEditFolder');

const client = new QueryClient({
  defaultOptions: {
    queries: {
      retry: false,
    },
  },
});

function ComponentFixture(props) {
  const nextProps = {
    canUpdate: true,
    ...props,
  };

  return (
    <QueryClientProvider client={client}>
      <IntlProvider locale="en" messages={{}}>
        <ThemeProvider theme={lightTheme}>
          <NotificationsProvider toggleNotification={() => {}}>
            <EditFolderDialog onClose={() => {}} {...nextProps} />
          </NotificationsProvider>
        </ThemeProvider>
      </IntlProvider>
    </QueryClientProvider>
  );
}

function setup(props = { onClose: jest.fn() }) {
  return render(<ComponentFixture {...props} />);
}

function getInput(container, name) {
  return container.querySelector(`input[name="${name}"]`);
}

function getButton(container, name) {
  return container.querySelector(`button[name="${name}"]`);
}

describe('EditFolderDialog', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  test('renders and matches the snapshot', () => {
    const { baseElement } = setup();
    expect(baseElement).toMatchSnapshot();
  });

  test('closes the modal', () => {
    const spy = jest.fn();
    const { baseElement } = setup({ onClose: spy });

    act(() => {
      fireEvent.click(getButton(baseElement, 'cancel'));
    });

    expect(spy).toBeCalledTimes(1);
  });

  test('name is a required field', async () => {
    const spy = jest.fn();
    const { baseElement } = setup({ onClose: spy });
    const name = getInput(baseElement, 'name');

    act(() => {
      fireEvent.click(getButton(baseElement, 'submit'));
    });

    expect(spy).toBeCalledTimes(0);

    act(() => {
      fireEvent.change(name, { target: { value: 'folder name' } });
    });

    act(() => {
      fireEvent.click(getButton(baseElement, 'submit'));
    });

    await waitFor(() => expect(spy).toBeCalledTimes(1));

    expect(spy).toBeCalledWith({ created: true });
  });

  test('set default form values', async () => {
    const spy = jest.fn();
    const folder = {
      id: 2,
      name: 'default folder name',
      createdAt: 'Mon Oct 24 2022 16:02:14 GMT+0200',
      updatedAt: 'Mon Oct 24 2022 16:02:14 GMT+0200',
      pathId: 2,
      path: '2',
      parent: null,
    };
    const { baseElement, queryByText } = setup({ folder, onClose: spy });

    expect(getInput(baseElement, 'name').value).toBe(folder.name);
    expect(queryByText('Media Library')).toBeInTheDocument();
  });

  test('set default form values with parentFolderId', async () => {
    const spy = jest.fn();
    const { queryByText } = setup({ parentFolderId: 2, onClose: spy });

    expect(queryByText('second child')).toBeInTheDocument();
  });

  test('show confirmation delete dialog', async () => {
    const folder = {
      id: 1,
      name: 'default folder name',
      createdAt: 'Mon Oct 24 2022 16:02:14 GMT+0200',
      updatedAt: 'Mon Oct 24 2022 16:02:14 GMT+0200',
      children: { count: 0 },
      parent: {
        id: 2,
        name: 'Some parent',
        createdAt: 'Mon Oct 24 2022 16:02:14 GMT+0200',
        updatedAt: 'Mon Oct 24 2022 16:02:14 GMT+0200',
        pathId: 2,
        path: '1/2',
      },
      pathId: 2,
      path: '2',
    };
    const folderStructure = [{ value: 1, label: 'Some parent' }];
    const { baseElement, queryByText } = setup({ folder, folderStructure });

    act(() => {
      fireEvent.click(getButton(baseElement, 'delete'));
    });

    expect(queryByText('Are you sure you want to delete this?')).toBeInTheDocument();
  });

  test('keeps edit folder dialog open and show error message on API error', async () => {
    const FIXTURE_ERROR_MESSAGE = 'folder cannot be moved inside itself';
    const moveSpy = jest.fn().mockRejectedValueOnce({
      response: {
        data: {
          error: {
            name: 'ValidationError',
            details: {
              errors: [
                {
                  path: ['parent'],
                  message: FIXTURE_ERROR_MESSAGE,
                  name: 'ValidationError',
                },
              ],
            },
          },
        },
      },
    });
    useEditFolder.mockReturnValueOnce({
      isLoading: false,
      editFolder: moveSpy,
    });

    const folder = {
      id: 1,
      name: 'default folder name',
      children: { count: 0 },
      createdAt: 'Mon Oct 24 2022 16:02:14 GMT+0200',
      updatedAt: 'Mon Oct 24 2022 16:02:14 GMT+0200',
      parent: {
        id: 2,
        name: 'Some parent',
        createdAt: 'Mon Oct 24 2022 16:02:14 GMT+0200',
        updatedAt: 'Mon Oct 24 2022 16:02:14 GMT+0200',
        pathId: 2,
        path: '1/2',
      },
      pathId: 2,
      path: '2',
    };
    const folderStructure = [{ value: 1, name: 'Some parent' }];

    const { getByText } = setup({ folder, folderStructure });

    const dialog = screen.getByRole('dialog');
    const submit = within(dialog).getByRole('button', {
      name: /save/i,
    });
    fireEvent.click(submit);

    await waitFor(() =>
      expect(moveSpy).toBeCalledWith({ name: 'default folder name', parent: null }, 1)
    );

    expect(getByText(FIXTURE_ERROR_MESSAGE)).toBeInTheDocument();
  });

  test('disables inputs and submit action if users do not have permissions to update', () => {
    useMediaLibraryPermissions.mockReturnValueOnce({ canUpdate: false });
    const { getByRole } = setup();

    expect(getByRole('textbox', { name: 'Name' })).toHaveAttribute('aria-disabled', 'true');
    expect(getByRole('combobox', { name: 'Location' })).toBeDisabled();

    expect(getByRole('button', { name: 'Create' })).toBeDisabled();
  });
});
