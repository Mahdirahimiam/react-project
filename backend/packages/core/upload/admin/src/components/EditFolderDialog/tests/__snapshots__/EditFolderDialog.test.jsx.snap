// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditFolderDialog renders and matches the snapshot 1`] = `
.c2 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

.c11 {
  font-size: 0.875rem;
  line-height: 1.43;
  font-weight: 600;
  color: #32324d;
}

.c20 {
  font-size: 0.75rem;
  line-height: 1.33;
  font-weight: 600;
  color: #32324d;
}

.c31 {
  font-size: 0.75rem;
  line-height: 1.33;
  font-weight: 600;
  color: #ffffff;
}

.c0 {
  margin-left: -250px;
  position: fixed;
  left: 50%;
  top: 2.875rem;
  z-index: 10;
  width: 31.25rem;
}

.c3 {
  padding-right: 40px;
  padding-left: 40px;
  position: fixed;
  z-index: 4;
}

.c6 {
  background: #ffffff;
  border-radius: 4px;
  box-shadow: 0px 2px 15px rgba(33,33,52,0.1);
  width: 51.875rem;
}

.c7 {
  background: #f6f6f9;
  padding-top: 16px;
  padding-right: 20px;
  padding-bottom: 16px;
  padding-left: 20px;
}

.c12 {
  background: #ffffff;
  padding: 8px;
  border-radius: 4px;
  border-color: #dcdce4;
  border: 1px solid #dcdce4;
  cursor: pointer;
}

.c15 {
  padding: 32px;
}

.c24 {
  padding-right: 12px;
}

.c29 {
  background: #4945ff;
  padding: 8px;
  padding-right: 16px;
  padding-left: 16px;
  border-radius: 4px;
  border-color: #4945ff;
  border: 1px solid #4945ff;
  cursor: pointer;
}

.c1 {
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 8px;
}

.c4 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c9 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c19 {
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 4px;
}

.c28 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  gap: 8px;
}

.c13 {
  position: relative;
  outline: none;
}

.c13 > svg {
  height: 12px;
  width: 12px;
}

.c13 > svg > g,
.c13 > svg path {
  fill: #ffffff;
}

.c13[aria-disabled='true'] {
  pointer-events: none;
}

.c13:after {
  -webkit-transition-property: all;
  transition-property: all;
  -webkit-transition-duration: 0.2s;
  transition-duration: 0.2s;
  border-radius: 8px;
  content: '';
  position: absolute;
  top: -4px;
  bottom: -4px;
  left: -4px;
  right: -4px;
  border: 2px solid transparent;
}

.c13:focus-visible {
  outline: none;
}

.c13:focus-visible:after {
  border-radius: 8px;
  content: '';
  position: absolute;
  top: -5px;
  bottom: -5px;
  left: -5px;
  right: -5px;
  border: 2px solid #4945ff;
}

.c30 {
  height: 2rem;
  border: 1px solid #dcdce4;
  background: #ffffff;
}

.c30 svg {
  height: 0.75rem;
  width: auto;
}

.c30[aria-disabled='true'] {
  border: 1px solid #dcdce4;
  background: #eaeaef;
}

.c30[aria-disabled='true'] .c10 {
  color: #666687;
}

.c30[aria-disabled='true'] svg > g,
.c30[aria-disabled='true'] svg path {
  fill: #666687;
}

.c30[aria-disabled='true']:active {
  border: 1px solid #dcdce4;
  background: #eaeaef;
}

.c30[aria-disabled='true']:active .c10 {
  color: #666687;
}

.c30[aria-disabled='true']:active svg > g,
.c30[aria-disabled='true']:active svg path {
  fill: #666687;
}

.c30:hover {
  background-color: #f6f6f9;
}

.c30:active {
  background-color: #eaeaef;
}

.c30 .c10 {
  color: #32324d;
}

.c30 svg > g,
.c30 svg path {
  fill: #32324d;
}

.c32 {
  height: 2rem;
}

.c32 svg {
  height: 0.75rem;
  width: auto;
}

.c32[aria-disabled='true'] {
  border: 1px solid #dcdce4;
  background: #eaeaef;
}

.c32[aria-disabled='true'] .c10 {
  color: #666687;
}

.c32[aria-disabled='true'] svg > g,
.c32[aria-disabled='true'] svg path {
  fill: #666687;
}

.c32[aria-disabled='true']:active {
  border: 1px solid #dcdce4;
  background: #eaeaef;
}

.c32[aria-disabled='true']:active .c10 {
  color: #666687;
}

.c32[aria-disabled='true']:active svg > g,
.c32[aria-disabled='true']:active svg path {
  fill: #666687;
}

.c32:hover {
  border: 1px solid #7b79ff;
  background: #7b79ff;
}

.c32:active {
  border: 1px solid #4945ff;
  background: #4945ff;
}

.c32 svg > g,
.c32 svg path {
  fill: #ffffff;
}

.c21 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c23 {
  border: none;
  border-radius: 4px;
  padding-bottom: 0.65625rem;
  padding-left: 16px;
  padding-right: 16px;
  padding-top: 0.65625rem;
  color: #32324d;
  font-weight: 400;
  font-size: 0.875rem;
  display: block;
  width: 100%;
  background: inherit;
}

.c23::-webkit-input-placeholder {
  color: #8e8ea9;
  opacity: 1;
}

.c23::-moz-placeholder {
  color: #8e8ea9;
  opacity: 1;
}

.c23:-ms-input-placeholder {
  color: #8e8ea9;
  opacity: 1;
}

.c23::placeholder {
  color: #8e8ea9;
  opacity: 1;
}

.c23[aria-disabled='true'] {
  color: inherit;
}

.c23:focus {
  outline: none;
  box-shadow: none;
}

.c22 {
  border: 1px solid #dcdce4;
  border-radius: 4px;
  background: #ffffff;
  outline: none;
  box-shadow: 0;
  -webkit-transition-property: border-color,box-shadow,fill;
  transition-property: border-color,box-shadow,fill;
  -webkit-transition-duration: 0.2s;
  transition-duration: 0.2s;
}

.c22:focus-within {
  border: 1px solid #4945ff;
  box-shadow: #4945ff 0px 0px 0px 2px;
}

.c17 {
  display: grid;
  grid-template-columns: repeat(12,1fr);
  gap: 16px;
}

.c18 {
  grid-column: span 6;
  max-width: 100%;
}

.c14 {
  border-color: #dcdce4;
  height: 2rem;
  width: 2rem;
}

.c14 svg g,
.c14 svg path {
  fill: #8e8ea9;
}

.c14:hover svg g,
.c14:focus svg g,
.c14:hover svg path,
.c14:focus svg path {
  fill: #666687;
}

.c14[aria-disabled='true'] svg path {
  fill: #666687;
}

.c5 {
  background: #32324d1F;
  inset: 0;
}

.c8 {
  border-radius: 4px 4px 0 0;
  border-bottom: 1px solid #eaeaef;
}

.c27 {
  border-radius: 0 0 4px 4px;
  border-top: 1px solid #eaeaef;
}

.c16 {
  overflow: auto;
  max-height: 60vh;
}

.c25 {
  background: transparent;
  border: none;
  position: relative;
  z-index: 1;
}

.c25 svg {
  height: 0.6875rem;
  width: 0.6875rem;
}

.c25 svg path {
  fill: #666687;
}

.c26 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background: none;
  border: none;
}

.c26 svg {
  width: 0.5625rem;
}

@media (max-width:68.75rem) {
  .c18 {
    grid-column: span;
  }
}

@media (max-width:34.375rem) {
  .c18 {
    grid-column: span 12;
  }
}

<body
  class="lock-body-scroll"
>
  <div>
    <div
      class="c0 c1"
    />
    <div
      class="c2"
    >
      <p
        aria-live="polite"
        aria-relevant="all"
        id="live-region-log"
        role="log"
      />
      <p
        aria-live="polite"
        aria-relevant="all"
        id="live-region-status"
        role="status"
      />
      <p
        aria-live="assertive"
        aria-relevant="all"
        id="live-region-alert"
        role="alert"
      />
    </div>
  </div>
  <div
    class=""
  >
    <div
      class="c3 c4 c5"
    >
      <div>
        <div>
          <div
            aria-labelledby="title"
            aria-modal="true"
            class="c6"
            role="dialog"
          >
            <form
              action="#"
              novalidate=""
            >
              <div
                class="c7 c8"
              >
                <div
                  class="c9"
                >
                  <h2
                    class="c10 c11"
                    id="title"
                  >
                    Add new folder
                  </h2>
                  <button
                    aria-disabled="false"
                    class="c12 c4 c13 c14"
                    type="button"
                  >
                    <span
                      class="c2"
                    >
                      Close the modal
                    </span>
                    <svg
                      aria-hidden="true"
                      fill="none"
                      focusable="false"
                      height="1rem"
                      viewBox="0 0 24 24"
                      width="1rem"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M24 2.417 21.583 0 12 9.583 2.417 0 0 2.417 9.583 12 0 21.583 2.417 24 12 14.417 21.583 24 24 21.583 14.417 12 24 2.417Z"
                        fill="#212134"
                      />
                    </svg>
                  </button>
                </div>
              </div>
              <div
                class="c15 c16"
              >
                <div
                  class="c17"
                >
                  <div
                    class="c18"
                  >
                    <div>
                      <div
                        class=""
                      >
                        <div
                          class="c19"
                        >
                          <label
                            class="c10 c20 c21"
                            for=":r0:"
                          >
                            Name
                          </label>
                          <div
                            class="c9 c22"
                          >
                            <input
                              aria-disabled="false"
                              aria-invalid="false"
                              aria-required="false"
                              class="c23"
                              id=":r0:"
                              name="name"
                              value=""
                            />
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    class="c18"
                  >
                    <div
                      class="c19"
                    >
                      <label
                        class="c10 c20 c21"
                        for="folder-parent"
                      >
                        Location
                      </label>
                      <div
                        class=" css-1iyvlxg-container"
                      >
                        <span
                          class="css-1f43avz-a11yText-A11yText"
                          id="react-select-2-live-region"
                        />
                        <span
                          aria-atomic="false"
                          aria-live="polite"
                          aria-relevant="additions text"
                          class="css-1f43avz-a11yText-A11yText"
                        />
                        <div
                          class=" css-1ioi85f-control"
                        >
                          <div
                            class=" css-onz4xi-ValueContainer"
                          >
                            <div
                              class=" css-1ryi7up-singleValue"
                            >
                              Media Library
                            </div>
                            <input
                              aria-autocomplete="list"
                              aria-expanded="false"
                              aria-haspopup="true"
                              aria-invalid="false"
                              aria-readonly="true"
                              class="css-mohuvp-dummyInput-DummyInput"
                              id="folder-parent"
                              inputmode="none"
                              role="combobox"
                              tabindex="0"
                              value=""
                            />
                          </div>
                          <div
                            class=" css-1hm8yhz-IndicatorsContainer"
                          >
                            <div
                              aria-hidden="true"
                              class="c24 c25 c26"
                            >
                              <svg
                                fill="none"
                                height="1rem"
                                viewBox="0 0 14 8"
                                width="1rem"
                                xmlns="http://www.w3.org/2000/svg"
                              >
                                <path
                                  clip-rule="evenodd"
                                  d="M14 .889a.86.86 0 0 1-.26.625L7.615 7.736A.834.834 0 0 1 7 8a.834.834 0 0 1-.615-.264L.26 1.514A.861.861 0 0 1 0 .889c0-.24.087-.45.26-.625A.834.834 0 0 1 .875 0h12.25c.237 0 .442.088.615.264a.86.86 0 0 1 .26.625Z"
                                  fill="#32324D"
                                  fill-rule="evenodd"
                                />
                              </svg>
                            </div>
                          </div>
                        </div>
                        <input
                          name="parent"
                          type="hidden"
                          value=""
                        />
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="c7 c27"
              >
                <div
                  class="c9"
                >
                  <div
                    class="c28"
                  >
                    <button
                      aria-disabled="false"
                      class="c29 c28 c13 c30"
                      name="cancel"
                      type="button"
                    >
                      <span
                        class="c10 c31"
                      >
                        Cancel
                      </span>
                    </button>
                  </div>
                  <div
                    class="c28"
                  >
                    <div
                      class="c28"
                    >
                      <button
                        aria-disabled="false"
                        class="c29 c28 c13 c32"
                        name="submit"
                        type="submit"
                      >
                        <span
                          class="c10 c31"
                        >
                          Create
                        </span>
                      </button>
                    </div>
                  </div>
                </div>
              </div>
            </form>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
